packages/pigeon/lib/cpp_generator.dart

```diff
@@ -800,79 +818,70 @@ const flutter::StandardMessageCodec& ${api.name}::GetCodec() {
-      final String argsSignature =
-          map2(argTypes, argNames, (String x, String y) => '$x $y').join(', ');
-      indent
-          .write('void ${api.name}::${func.name}($argsSignature, $callback) ');
-    }
+    final HostDatatype returnType = getHostDatatype(
+        func.returnType,
+        root.classes,
+        root.enums,
+        (TypeDeclaration x) => _baseCppTypeForBuiltinDartType(x));
+    final String successCallback =
+        'std::function<void(${_flutterApiReturnType(returnType)})>&& on_success';
+    const String errorCallback =
+        'std::function<void(const FlutterError&)>&& on_error';
+
+    // Determine the input argument list, saved in a structured form for later
+    // use as platform channel call arguments.
+    final Iterable<_HostNamedType> hostArgs =
+        indexMap(func.arguments, (int i, NamedType arg) {
+      final HostDatatype hostType = getFieldHostDatatype(arg, root.classes,
+          root.enums, (TypeDeclaration x) => _baseCppTypeForBuiltinDartType(x));
+      return _HostNamedType(_getSafeArgumentName(i, arg), hostType, arg.type);
+    });
+    final String inputArgumentList = hostArgs
+        .map((_HostNamedType arg) =>
+            '${_flutterApiArgumentType(arg.hostType)} ${arg.name}')
+        .join(', ');
+    indent.write('void ${api.name}::${_makeMethodName(func)}('
+        '${inputArgumentList.isEmpty ? '' : '$inputArgumentList, '}'
+        '$successCallback, $errorCallback) ');
     indent.scoped('{', '}', () {
       const String channel = 'channel';
       indent.writeln(
-          'auto channel = std::make_unique<flutter::BasicMessageChannel<flutter::EncodableValue>>(');
-      indent.inc();
-      indent.inc();
-      indent.writeln('binary_messenger_, "$channelName", &GetCodec());');
-      indent.dec();
-      indent.dec();
+          'auto channel = std::make_unique<flutter::BasicMessageChannel<>>(binary_messenger_, '
+          '"$channelName", &GetCodec());');
+
+      // Convert arguments to EncodableValue versions
+      const String argumentListVariableName = 'encoded_api_arguments';
+      indent.write('flutter::EncodableValue $argumentListVariableName = ');
+      if (func.arguments.isEmpty) {
+        indent.addln('flutter::EncodableValue();');
+      } else {
+        indent.scoped('flutter::EncodableValue(flutter::EncodableList{', '});',
+            () {
+          for (final _HostNamedType arg in hostArgs) {
+            final String encodedArgument = _wrappedHostApiArgumentExpression(
+                root, arg.name, arg.originalType, arg.hostType);
+            indent.writeln('$encodedArgument,');
+          }
+        });
+      }
+
       indent.write(
-          '$channel->Send($sendArgument, [callback](const uint8_t* reply, size_t reply_size) ');
+          '$channel->Send($argumentListVariableName, [on_success, on_error](const uint8_t* reply, size_t reply_size) ');
```

Comment:
nit: We can move these callbacks instead of copying them.  Who knows what clients will have in them.  If it is a pain, feel free to leave for later with a note.

Reply:
Done; I tried to do this with `reply` before realizing we need to keep it for fallback error handling, and I just forgot to do it here.